### PR TITLE
Stash

### PR DESCRIPTION
**Pull request by qndel**
image

Atm turned on/off using X
All characters see the same stash

Issues:
~~Item placement somehow seems different than in inventory - you have to aim carefully to put item in the slot you want~~ (Thanks @FluffyQuack ;))

TODO:
~~Add item descriptions~~
Add more tabs
Put the stash to some npc or an actual chest in town
~~Close the stash when you leave the game~~

### CODE REVIEW

**Comment by sheepo99** [comment]
Again, I suggest the stash should be accessed by talking to Ogden.

**Comment by sheepo99** [comment]
For reference, this is how Beelzebub mod handled tab design.

Diablo-1-HD-06

**Comment by joewis** [comment]
This is awesome. Just hoping for this to be added to 1.2.0 milestone :)
I lol'd a bit at the filename.

Would it be possible to use the Questlog background instead of the Inventory background, so there won't be the half head at the top?

**Comment by FitzRoyX** [comment]
> 
> Would it be possible to use the Questlog background instead of the Inventory background, so there won't be the half head at the top?

Pretty sure it's going to be covered up by tabs anyway.

**Comment by qndel** [comment]
> This is awesome. Just hoping for this to be added to 1.2.0 milestone :)
> I lol'd a bit at the filename.
> 
> Would it be possible to use the Questlog background instead of the Inventory background, so there won't be the half head at the top?

Quest background's width is 1 pixel different than inventory so frames won't match

**Comment by sheepo99** [comment]
@qndel what about using a custom graphics frame?

**Comment by qndel** [comment]
> @qndel what about using a custom graphics frame?

I don't have one atm

**Comment by galaxyhaxz** [comment]
Credits must be given UwU

inv

**Comment by qndel** [comment]
ye because grid was what I'm missing ;)

**Comment by sheepo99** [comment]
@qndel apparently there's someone making custom UI stuff for the mobile ports on #493. Maybe you can ask that person for help?

**Comment by FitzRoyX** [comment]
> @qndel apparently there's someone making custom UI stuff for the mobile ports on #493. Maybe you can ask that person for help?

There's probably a way to do it with the existing graphics assets, which is preferable.

**Comment by galaxyhaxz** [comment]
Excuse my MS Paint skills :P
test

**Comment by sheepo99** [comment]
This is lovely, but we might actually need a smaller grid, because we need to make room for a close window button, along with the tabs.

**Comment by qndel** [comment]
> This is lovely, but we might actually need a smaller grid, because we need to make room for a close window button, along with the tabs.
close window? wtf

**Comment by sheepo99** [comment]
Oh wait, you're right, we can turn it off just by pressing the Inv button.

**Comment by galaxyhaxz** [comment]
Yeah, or escape. D2 had manual close buttons for everything, personally i'm not a fan of them.

**Comment by sheepo99** [comment]
Tbh, close buttons might still be needed for widescreen mode though. I'd assume it's a bit more comfortable not having to scroll your mouse all the way down.

**Comment by FitzRoyX** [comment]
Close buttons don't have to be large or graphical, look at what beelzebub does for message windows, it's just an overlaid x.

**Comment by galaxyhaxz** [comment]
> Close buttons don't have to be large or graphical, look at what beelzebub does for message windows, it's just an overlaid x.

Yeah, I think they'd be required for a D2 like interface for vendors, since you don't have a button to close out of it.

File: SourceX/stash.cpp [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+		fclose(pFile);
+		return false;
+	}
+	/* the whole file is now loaded in the memory buffer. */
+	// terminate
+	fclose(pFile);
+
+	tbuff = tmp;
+	CopyInt(tbuff, &_pNumStash);
+	LoadItems(NUM_STASH_GRID_ELEM, StashList);
```
Comment (AJenbo) [inline SourceX/stash.cpp:159]: Don't forget to set gbIsHellfireSavegame before calling the savegame functions.

Comment (qndel) [reply SourceX/stash.cpp:159]: isn't it set already?

Comment (AJenbo) [reply SourceX/stash.cpp:159]: It gets set to different values depending on what operation is about to be done. So you need to set it to the appropriate value before calling any of the loading routines to be safe.

reading hellfire save = true
reading diablo save = false
networking diablo mode = false
networking hellfire mode = true
cornerstone = true

```suggestion
	gbIsHellfireSaveGame = gbIsHellfire;
	LoadItems(NUM_STASH_GRID_ELEM, StashList);
```

**Comment by NikoVP** [comment]
Hi, A useful feature of the stash would be to have withdraw button for gold. where the dialog for splitting gold piles would be used to enter the amount desired to be taken. Would be nice to have some things in mind such as available space in the inventory and automatic movement and placement of gold in the first tab where gold pile has been detected, this will keep inventory easier to organize. And logic to move to next available tab when gold pile overflows the limit. I know this complicates the logic but will save time for gold management. Which could become a problem in the long run. Also buying items with gold from the stash would also be a time saver. Cheers!

**Comment by AJenbo** [comment]
could just hold gold as a combined total and not have any piles, isnt that how the stash worked in d2?

**Comment by NikoVP** [comment]
In that case what would a reasonable maximum for stashed gold be, before abusers fill it cloned/hacked gold?

**Comment by AJenbo** [comment]
1,250,000, but this wont stop people from cheating if they know how to hack

**Comment by qndel** [comment]
I'm not going to handle gold in any special way ;)

**Comment by FitzRoyX** [comment]
Going fullblown D2 with gold seems unnecessary. Beelzebub just raised the 5000 cap to 1,000,000 and it was fine. All the best items are found anyway, accumulating massive amounts of gold doesn't really break anything.

**Comment by galaxyhaxz** [comment]
> could just hold gold as a combined total and not have any piles, isnt that how the stash worked in d2?

Maybe have it work in a way similar to Darkstone instead. In Darkstone, there is a bank teller in town and you can talk to him and see how much gold you have in your account. You can deposit all the gold, or withdraw and it automatically creates the piles in inventory. Gold is still useful in D1, and I didn't like having to drop piles all over town in case Griswold had a good item stock.

Let's be Frank, and Frank is a pretty cool guy except being turned evil from Neural transmitters. Once gold goes into the stash there is literally never any reason to take it out, except giving to another player.

DIABLO_20210303_035148

**Comment by FitzRoyX** [comment]
The question is, is it any faster? Moving a giant pile of gold to and from your stash is pretty fast. That's probably why bub never bothered to have separate banking options.

**Comment by galaxyhaxz** [comment]
It was just an idea tossed out there, either method works ^^

**Comment by Tremain84** [comment]
Sorry for the question. How can I open the stash Ingame ? I tried the "x" button and many more. Or is this a feature not implemented yet?

PS: I also think stacking gold and holding the mouse key to attack is not changing the game experience but enhancing it.

**Comment by AJenbo** [comment]
> Sorry for the question. How can I open the stash Ingame ? I tried the "x" button and many more. Or is this a feature not implemented yet?
> 
> PS: I also think stacking gold and holding the mouse key to attack is not changing the game experience but enhancing it.

What version are you testing?

See https://github.com/diasurgical/devilutionX/pull/895 for the status on holding the mouse to attack.

**Comment by qndel** [comment]
as you can see, this pull request is not merged yet, means you'd have to compile it to use the stash but I expect it to make it to 1.2.0 version

**Comment by Tremain84** [comment]
> 
> 
> > Sorry for the question. How can I open the stash Ingame ? I tried the "x" button and many more. Or is this a feature not implemented yet?
> > PS: I also think stacking gold and holding the mouse key to attack is not changing the game experience but enhancing it.
> 
> What version are you testing?
> 
> See #895 for the status on holding the mouse to attack.

I am running the latest deviloutionX release 1.1.0.
Alright, so if I want more quality of life features I need to install additional patches like the compiled QOL Patches on top of the 1.1.0 version.


@qndel Thank you so much for the project and your work. I am a die hard D1 + D2 fan. I played both games upon release till today (not a hellfire fan). Love the new widescreen!
Ok, then I wait for the 1.2.0 version in order to experience a stash and higher gold stack, which are besides the hold mouse to attack the most important QOL features to add in my opinion.

**Comment by qndel** [comment]
https://discord.gg/9eCtSCz244
Feel free to join our discord server, I could probably cook you up an exe with all the cool features on monday ;)

**Comment by AJenbo** [comment]
>  Ok, then I wait for the 1.2.0 version in order to experience a stash and higher gold stack

There are no plans for higher gold stacks in 1.2.0

**Comment by Tremain84** [comment]
> 
> 
> https://discord.gg/9eCtSCz244
> Feel free to join our discord server, I could probably cook you up an exe with all the cool features on monday ;)

Thanks, I just joined the discord channel (Tremain84).
Ok that would be really amazing. I would love to have just some QOL changes to keep it very vanilla, like:

**Shareable stash between the characters.**
Increased gold count stack.
Hold mousebutton to attack.
Saving hotkeys.
Highlight items on the ground when hitting alt (finding rings can otherwise be a pain :D

**Not needed but nice to also have included:**
Player experience bar.
Item values shown when mouse is over an item in inventory
Auto gold pick up

**Future projects ?**
Adding floating damage numbers.
Moving Wirt closer to town, so trying to gamble the kings sword of haste/speed doesn't take ages ;)
A drag and drop buy/sell window while at a vendor (I know this is a big one).
After killing Diablo beeing able to pick up the loot dropped.
Including the high framerate mod (increases the fps from 25 to 60 fps.)

PS: maybe offtopic, but is there a support/donation button at gitgub? Or is there any other way I can support this project?

**Comment by AJenbo** [comment]
> **Not needed but nice to also have included:**
> Player experience bar.
> Auto gold pick up

Already part of the betas.

> Including the high framerate mod (increases the fps from 25 to 60 fps.)

D2 is 25, D1 is 20. We already render at monitor refresh rate (so 60+), but positions are still updated at just 20fps (or whatever you set the game speed to).

> PS: maybe offtopic, but is there a support/donation button at gitgub? Or is there any other way I can support this project?

The topic you are writing in is about Stack, everything else is off-topic.

We do not accept donations. Play the game, have fun, spread the word be nice.

**Comment by Tremain84** [comment]
> 
> 
> > **Not needed but nice to also have included:**
> > Player experience bar.
> > Auto gold pick up
> 
> Already part of the betas.
> 
> > Including the high framerate mod (increases the fps from 25 to 60 fps.)
> 
> D2 is 25, D1 is 20. We already render at monitor refresh rate (so 60+), but positions are still updated at just 20fps (or whatever you set the game speed to).
> 
> > PS: maybe offtopic, but is there a support/donation button at gitgub? Or is there any other way I can support this project?
> 
> The topic you are writing in is about Stack, everything else is off-topic.
> 
> We do not accept donations. Play the game, have fun, spread the word be nice.

Ok :)
Thank you very much. If you can manage to create that exe including the QOL changes I would be super happy :)
Keep up the great work and I will be on here a lot more, just need to get used to the github layout.

**Review by Josegto1 — APPROVED** [review]

**Comment by qndel** [comment]
@Josegto1 I'd appreciate if you stopped "approving" random reviews without having 0 clue about what's going on - you reviewed 33 PRs in random repos today and all of them had literally 0 value, everything ok with your head?

**Review by Josegto1 — APPROVED** [review]

**Comment by Tremain84** [comment]
> 
> 
> > Ok, then I wait for the 1.2.0 version in order to experience a stash and higher gold stack
> 
> There are no plans for higher gold stacks in 1.2.0

Well you are right, if we have a stash in version 1.2.0 then a higher gold stack is not required, because we can store gold in the stash. That makes sense :)
atm I have to throw a lot of gold away at the end of each session, because I can't carry all of it.

**Comment by ikonomov** [comment]
I really hope town-only accessible stashes eventually make it into the game, two separate ones for multi and single player.

**Comment by qndel** [comment]
Please keep your hopes away from old PRs.

**Pull request closed without merging**